content/shmem_malloc_hints.tex

```diff
@@ -0,0 +1,97 @@
+    \VAR{size} bytes, which shall be suitably aligned so that it may be
+    assigned to a pointer to any type of object.  This space is allocated from
+    the symmetric heap (similar to \FUNC{shmem\_malloc}).  When \VAR{size} is zero, 
+    the \FUNC{shmem\_malloc\_with\_hint} routine performs no action and returns a null pointer. 
+    
+    In addition to the \VAR{size} argument, the \VAR{hint} argument is provided by the user. 
+    The \VAR{hint} describes the expected manner in which the \openshmem program may use the allocated memory.
+    The valid usage hints are described in Table~\ref{usagehints}. Multiple hints are expressed as \CONST{OR} of \VAR{hints}.
+    
+    The information provided by the \VAR{hint} is used to optimize for performance by the implementation. 
+    If the implementation cannot optimize, the behavior is same as \FUNC{shmem\_malloc}.
+    If more than one hint is provided, the implementation will make the best effort to use one or more hints 
+    to optimize performance. 
+            
+    The \FUNC{shmem\_malloc\_with\_hint} routine is provided  so that multiple \acp{PE} in a program can allocate symmetric,
+    remotely accessible memory blocks.  When no action is performed, these
+    routines return without performing a barrier. Otherwise, the routine will call a barrier on exit.
+    This ensures that all \acp{PE} participate in the memory allocation, and that the memory on other
+    \acp{PE} can be used as soon as the local \ac{PE} returns. The implicit barrier performed by this routine will quiet the
+    default context.  It is the user's responsibility to ensure that no communication operations involving the given memory block are pending on
+    other contexts prior to calling the \FUNC{shmem\_free} and \FUNC{shmem\_realloc} routines.
+    The user is also responsible for calling these routines with identical argument(s) on all
+    \acp{PE}; if differing \VAR{size}, or \VAR{hint} arguments are used, the behavior of the call
+    and any subsequent \openshmem calls is undefined.
+}
+
+\apireturnvalues{
+    The \FUNC{shmem\_malloc\_with\_hint} routine returns a pointer to the allocated space;
+    otherwise, it returns a null pointer.
+}
+
+\apinotes{ 
+}
+
+\apiimpnotes{
+}
+	\begin{longtable}{|p{0.45\textwidth}|p{0.5\textwidth}|}
+	\hline
+	\textbf{Hints} & \textbf{Usage hint}
+	\tabularnewline \hline
+	\endhead
+	%%
+	\LibConstDecl{SHMEM\_HINT\_NONE} &
+	Behavior same as shmem\_malloc
+	\tabularnewline \hline
+		
+	\LibConstDecl{SHMEM\_HINT\_LOW\_LAT\_MEM} &
+	Allocate memory on low-latency storage	
+	\tabularnewline \hline
+	
+	\LibConstDecl{SHMEM\_HINT\_HIGH\_BW\_MEM} &
+	Allocate memory on high-bandwidth storage	
+	\tabularnewline \hline
+
+	\LibConstDecl{SHMEM\_HINT\_PSYNC} &
+	Memory used as \CONST{PSYNC} array
+	\tabularnewline \hline
+		
+	\LibConstDecl{SHMEM\_HINT\_PWORK} &
+	Memory used as \CONST{PWORK} array
+	\tabularnewline \hline
+	
+	\LibConstDecl{SHMEM\_HINT\_ATOMICS} &
```

Comment:
What happens with ATOMIC hint was not passed but atomic operation was used ?

Reply:
I don't explicitly cover the bad behavior semantics. But, based on the use case experience we have currently, I'm inclined to say that It should not impact correctness. However, the programs might not be able to get the best performance characteristics for that environment.

Do you prefer that we should explicitly say this?

Reply:
My initial assumption is that if this is hint, implementation can completely ignore it. But we still have issue with this semantics. For HCA optimal location of the memory for Atomics will be on the HCA. For core it will be optimal to have in L1/L2 which is suboptimal for ATOMICs coming from the network and vice-versa. Essentially such hint may actually do more  harm to applications. My suggestion is to use explicit names - HCA1, HCA2, GPU, L2, LLC, etc. This will provide very clear description to user what actually happens.

Reply:
Since in this case you optimize for remote PEs only, I think it should be named SHMEM\_HINT\_REMOTE\_ATOMICS to be clear that you don't optimize for local PEs. Without explicit specification local/remote the hint is not very useful.